(GH-2076) Update 'Writing plugins' docs page (#2250)
puppetlabs/bolt: documentation/writing_plugins.md

# Writing plugins

A plugin is a special task that Bolt loads from a module on each run. Before you
write a plugin, get familiar with [writing tasks](writing_tasks.md).

## Supported languages

Bolt runs plugin tasks on the `localhost` target without any configuration from
the inventory file. You can write your plugin in any scripting language, as long
as your workstation can interpret that scripting language.

> For Unix-like controllers, you should include a shebang line at the top of of

Reviewer: ```suggestion
For Unix-like controllers, you should include a shebang line at the top of
```

├── metadata.json
└── tasks
    ├── resolve_reference.py
    └── resolve_reference.json
```

### Reference plugins

Reference plugins fetch data from an external source and store it in a static
data object. For example, you might use a reference plugin to fill in the
password field of a configuration file with the contents of an environment
variable, or to query AWS for a list of targets to populate your inventory file.

There are two plugin hooks associated with reference plugins:

- `resolve_reference`

  This hook is used by the reference plugin itself. Most plugins use this hook.

- `validate_resolve_reference`

  This hook is used to pre-validate the parameters that will be passed to the
  `resolve_reference` plugin. These plugins let Bolt raise any validation
  errors during inventory loading, rather than in the middle of a plan run.
  Regardless of whether this hook is specified, Bolt tests parameters to
  make sure they match the parameters for the `resolve_reference` plugin.

📖 **Related information**

- For an example of a reference plugin, see the built-in [yaml
  plugin](https://github.com/puppetlabs/puppetlabs-yaml/tree/master/tasks).

### Secret plugins

> Use a secret plugin to create keys for encryption and decryption, encrypt
> plaintext, and decrypt ciphertext. You can configure Bolt to use specific secret
> plugins when you run the `bolt secret` command.

Reviewer: ```suggestion
Use a secret plugin to create keys for encryption and decryption, to encrypt
plaintext, or to decrypt ciphertext. You can configure Bolt to use specific secret
plugins when you run the `bolt secret` command.
```

# Writing plugins

Plugins are written as tasks that are loaded from modules when you run Bolt. If
you are familiar with writing tasks, then you can write plugins! If you are not
familiar with how to write a task, we recommend that you read more about
[writing tasks](writing_tasks.md) before continuing on.

## Supported languages

Because plugins are written as tasks, they can be written in any scripting
language supported by the target it runs on. Plugins are always run on the
`localhost` target without any configuration from the inventory file, so as long
as your controller can interpret the scripting language the plugin is written

> with, you can run the plugin.

Author: The controller is the workstation/system/whatever that is running Bolt. It's a term we use elsewhere, though I do think workstation is clearer.

For Unix-like controllers, you should include a shebang line at the top of of
the plugin file to ensure it is executed using the correct interpreter.

```python
#!/usr/bin/env python

import json, os, sys

...
```

## Module structure

Because plugins are written as tasks, they are part of a module. A simple

> module including a plugin will have a structure similar to this:

Author: Ah, that should definitely be a `my_plugin` module directory with a `resolve_reference` plugin. I was originally trying to avoid introducing the concept of hooks too early, but it just adds more confusion.

give the task a name matching the name of the hook. For example, to create a
plugin named `read_yaml` that uses the `resolve_reference` hook, your module
would look like this:

```shell
read_yaml/
├── bolt_plugin.json
├── metadata.json
└── tasks
    ├── resolve_reference.py
    └── resolve_reference.json
```

### Reference plugins

Reference plugins fetch data from an external source and store it in a static
data object. For example, a reference plugin may read an environment variable's
value and use it to fill in the password field of a configuration file, while
another reference plugin may query AWS for a list of targets to populate the
inventory with.

There are two plugin hooks associated with reference plugins:

- `resolve_reference`

  This hook is used by the reference plugin itself. Most plugins will use this
  hook.

- `validate_resolve_reference`

  This hook is used to pre-validate the parameters that will be passed to the
  `resolve_reference` plugin. These plugins let Bolt raise any validation
  errors during inventory loading, rather than in the middle of a plan run.
  Regardless of whether this hook is specified, Bolt tests parameters to

>   make sure they match the parameters for the `resolve_reference` plugin.

Author: It allows you to do some additional validation that can't be done with just metadata, like making sure you can actually connect to a service or that credentials exist at a specified location on the workstation. Most users won't use them, and I don't think we necessarily want to push their use at this time.

>   make sure they match the parameters for the `resolve_reference` plugin.

Reviewer: Okay, no worries. If we don't want to expand on those uses then I think we're good here

📖 **Related information**

- [Example: yaml
  plugin](https://github.com/puppetlabs/puppetlabs-yaml/tree/master/tasks)

### Secret plugins

> Secret plugins are used to create keys for encryption and decryption, encrypt
> plaintext, and decrypt ciphertext. Bolt can be configured to use specific
> secret plugins when it runs the `bolt secret` command.

Reviewer: ```suggestion
Use a secret plugin to create keys for encryption and decryption, encrypt
plaintext, and decrypt ciphertext. You can configure Bolt to use specific
secret plugins when you run the `bolt secret` command.
```

There are three plugin hooks associated with secret plugins:

- `secret_createkeys`

>   This hook is used for a plugin that creates keys for encryption and
>   decryption. Bolt always passes a `force` parameter to plugins using this hook.
>   When running the `bolt secret createkeys --force` command or
>   `New-BoltSecretKey -Force` cmdlet, the `force` parameter will have a value of
>   `true`.

Reviewer: ```suggestion
  Use the `secret_createkeys` hook for plugins that create keys for encryption and
  decryption. Bolt always passes a `force` parameter to plugins using this hook.
  When running the `bolt secret createkeys --force` command or
  `New-BoltSecretKey -Force` cmdlet, the `force` parameter has a value of
  `true`.
```

- `secret_decrypt`

>   This hook is used for a plugin that decrypts ciphertext and returns plaintext.

Reviewer: ```suggestion
  Use the `secret_decrypt` hook for plugins that decrypt ciphertext and return plaintext.
```

>   Bolt always passes an `encrypted_value` parameter to plugins using this hook,
>   which is set by running the `bolt secret decrypt` command or
>   `Unprotect-BoltSecret` cmdlet.

Reviewer: ```suggestion
  Bolt always passes an `encrypted_value` parameter to plugins using this hook.
  The `encrypted_value` parameter is set when you run the `bolt secret decrypt` 
  command or `Unprotect-BoltSecret` cmdlet.
```

- `secret_encrypt`

>   This hook is used for a plugin that encrypts plaintext and returns ciphertext.
>   Bolt always passes a `plaintext_value` parameter to plugins using this hook,
>   which is set by running the `bolt secret encrypt` command or
>   `Protect-BoltSecret` cmdlet.

Reviewer: ```suggestion
  Use the `secret_encrypt` hook for plugins that encrypt plaintext and return 
  ciphertext. Bolt always passes a `plaintext_value` parameter to plugins using 
  this hook. The `plaintext_value` parameter is set when you run the 
  `bolt secret encrypt` command or `Protect-BoltSecret` cmdlet.
```

📖 **Related information**

> - [Example: pkcs7
>   plugin](https://github.com/puppetlabs/puppetlabs-pkcs7/tree/master/tasks)

Reviewer: ```suggestion
- For an example of a secret plugin, see the built-in [pkcs7
  plugin](https://github.com/puppetlabs/puppetlabs-pkcs7/tree/master/tasks).
```

### Puppet library plugins

> Bolt uses a Puppet library plugin to make sure the Puppet library is available
> on a target when the `apply_prep` function is called.

Reviewer: ```suggestion
Puppet library plugins install Puppet libraries on a target when a plan calls
the `apply_prep` function.
```

> Puppet library plugins behave differently from other types of plugins. While
> other types of plugins are run on `localhost`, Puppet library plugins are run on
> each target that you are running `apply_prep` on. Because of this, each target
> you run the plugin on must be able to interpret the scripting language the
> plugin uses.

Reviewer: ```suggestion
Puppet library plugins behave differently from other types of plugins. While
Bolt runs other types of plugins on `localhost`, Bolt runs Puppet library 
plugins on each target that you are running `apply_prep` on. 
Because of this behavior, each target you run the plugin on must be able 
to interpret the scripting language the plugin uses.
```

There is a single plugin hook associated with Puppet library plugins:

- `puppet_library`

>   This hook is used for plugins that make sure the Puppet library is
>   available on a target.

Reviewer: ```suggestion
  Use the `puppet_library` hook for plugins that make sure the Puppet library is
  available on a target.
```

📖 **Related information**

> - [Example: puppet_agent
>   plugin](https://github.com/puppetlabs/puppetlabs-puppet_agent/tree/master/tasks)

Reviewer: ```suggestion
- For an example of a Puppet library plugin, see the built-in [puppet_agent
  plugin](https://github.com/puppetlabs/puppetlabs-puppet_agent/tree/master/tasks).
```

## Configuring hooks

> The `bolt_plugin.json` not only indicates to Bolt that the module includes a
> plugin, but also allows you to map tasks to specific plugin hooks. Mapping a
> task to a plugin hook allows you to give the task any valid name you want while
> still having Bolt recognize the task as a plugin. You can also configure
> multiple hooks for a single plugin.

Reviewer: ```suggestion
The `bolt_plugin.json` file not only indicates to Bolt that the module includes a
plugin, but also allows you to map tasks to specific plugin hooks. Mapping a
task to a plugin hook allows you to give the task any valid name you want while
still having Bolt recognize the task as a plugin. You can also use the 
`bolt_plugin.json` file to configure multiple hooks for a single plugin.
```

  "hooks": {
    "resolve_reference": {
      "task": "my_module::my_plugin"
    }
  }
}
```

Plugins can also support multiple hooks. For example, you might want to use
a `secret_decrypt` plugin as a `resolve_reference` plugin:

```json
{
  "hooks": {
    "resolve_reference": {
      "task": "pkcs7::secret_decrypt"
    }
  }
}
```

## Plugin input

Plugins can accept structured input. For example, a plugin that retrieves a list
of targets from a service may need credentials to authenticate with the service,
while a plugin that decrypts a value may need a path to a key pair.

Because plugins are written as tasks, you can pass input to a plugin by defining
parameters in the task metadata file and then specifying the parameters when you
use the plugin. A simple plugin that loads a YAML file may include a single
`filepath` parameter in the task metadata:

```json
{
  "description": "Read YAML data from a file.",
  "input_method": "stdin",
  "parameters": {
    "filepath": {
      "type": "String",
      "description": "The path to the YAML file."
    }
  }
}
```

Depending on the `input_method` defined in the task metadata, you can access the
plugin's input in a few different ways.

> - `stdin`: Read the input parameters from `stdin` and parse as JSON

Reviewer: ```suggestion
- `"stdin"`: Read the input parameters from standard input (stdin) and parse as JSON.
```

> - `environment`: Read the input parameters from environment variables matching
>   the parameter's name and prefixed with `PT_`, for example `PT_filepath`

Reviewer: ```suggestion
- `"environment"`: Read the input parameters from environment variables matching
  the parameters' names and prefixed with `PT_`, for example `PT_filepath`.
```

> - `powershell`: Read the input parameters from named arguments matching the
>   parameter's name

Reviewer: ```suggestion
- `"powershell"`: Read the input parameters from named arguments matching the
  parameters' names
```

> Because the task metadata above uses the `stdin` input method, the task will
> need to read parameters from `stdin` and parse the input as JSON:

Reviewer: ```suggestion
Because the task metadata from the YAML plugin above uses the stdin input method, 
the task needs to read parameters from stdin and parse the input as JSON:
```

```python
#!/usr/bin/env python
import json, sys, yaml

params = json.load(sys.stdin)
filepath = params['filepath']

...
```

> If the task metadata above instead used the `environment` input method, the
> parameter could be read from the `PT_filepath` environment variable:

Reviewer: ```suggestion
If the task metadata used the environment input method instead of stdin, 
the task could read the parameter from the `PT_filepath` environment variable:
```

```python
#!/usr/bin/env python
import json, os, sys, yaml

filepath = os.environ['PT_filepath']

...
```

### Metaparameters

> Because plugins are written as tasks, all plugins receive metaparameters from
> Bolt by default. These metaparameters are helpful if your plugin task uses other
> files are needs to locate files relative to the Bolt project directory. The
> following metaparameters are available:

Reviewer: ```suggestion
Like all tasks, plugins receive metaparameters from Bolt by default. 
These metaparameters are helpful if your plugin task uses other
files or needs to locate files relative to the Bolt project directory. The
following metaparameters are available:
```
Not sure about that `or`.

- `_boltdir`

  The absolute path to the Bolt project directory. This is useful when you need
  to expand a path relative to the Bolt project directory.

- `_installdir`

  The temporary directory that the task is installed to. This is useful when the
  task uses additional files specified in the task metadata.

📖 **Related information**

- [Writing tasks: Using structured input and
  output](writing_tasks.md#using-structured-input-and-output)

## Plugin output

Plugins should return output. The format of the plugin output depends on how you
are using the plugin. For example, a plugin that encrypts a string should return
an encrypted string, while a plugin that looks up a list of targets from a
service should return structured data similar to what would be written in an
inventory file.

As a plugin author, it's important that you consider the use case for your

> plugin and the data it is expected to output. For example, we might want to

Reviewer: ```suggestion
plugin and the data it is expected to output. For example, you might want to
```

The output from the plugin may look similar to this:

```json
{
  "value": [
    {
      "name": "windows",
      "config": {
        "transport": "winrm"
      },
      "targets": [
        "https://win-1.example.com",
        "https://win-2.example.com"
      ]
    }
  ]
}
```

Which would result in an inventory that looks like this:

```yaml
groups:
  - name: windows
    config:
      transport: winrm
    targets:
      - https://win-1.example.com
      - https://win-2.example.com
```

📖 **Related information**

- [Writing tasks: Using structured input and
  output](writing_tasks.md#using-structured-input-and-output)

## Returning errors

For all but the most simple plugins, it can be helpful to validate input, handle
exceptions, and generally ensure that nothing goes wrong during the execution of
the plugin. If something does go wrong, your plugin should return an error
object, which is the standard way of returning errors from tasks.

Error objects include a single `_error` key, which accepts an object that must
include a `msg` key. For example, a minimal error object may look like this:

```json
{
  "_error": {
    "msg": "Something went horribly, horribly wrong."
  }
}
```

> This error object would be parsed by Bolt, which would then raise an error
> similar to this:

Reviewer: ```suggestion
Bolt would parse this error object and raise an error similar to this:
```

```shell
Error executing plugin example from resolve_reference in example:

> Something went horribly, horribly wrong.

Reviewer: 😂
